src/webgpu/shader/execution/expression/binary/bitwise_shift.spec.ts

```diff
@@ -0,0 +1,311 @@
+    const must_be_zero_msb = e2;
+    const mask = ~0 << (bitwidth - must_be_zero_msb);
+    if ((e1 & mask) !== 0) {
+      return false;
+    }
+  } else {
+    // If T is a signed integer type, and the e2+1 most significant bits of e1 do
+    // not have the same bit value, then error.
+    const must_match_msb = e2 + 1;
+    const mask = ~0 << (bitwidth - must_match_msb);
+    if ((e1 & mask) !== 0 && (e1 & mask) !== mask) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Returns true if e1 >> e2 is valid for const evaluation
+function is_valid_const_shift_right(e1: number, e1Type: string, e2: number) {
+  // Shift by 0 is always valid
+  if (e2 === 0) {
+    return true;
+  }
+
+  // Cannot shift by bitwidth or greater
+  if (e2 >= bitwidth) {
+    return false;
+  }
+
+  return true;
+}
+
+// Returns all cases of shifting e1 left by [0,63]. If `is_const` is true, cases that are
+// invalid for const eval are not returned.
+function generate_shift_left_cases(e1: number, e1Type: string, is_const: boolean): CaseList {
+  const V = e1Type === 'i32' ? i32 : u32;
+  const cases: CaseList = [];
+  for (let shift = 0; shift < 64; ++shift) {
+    const e2 = shift;
+    if (is_const && !is_valid_const_shift_left(e1, e1Type, e2)) {
+      // console.log(`Skipping: ${e1} << ${e2}`);
+      continue;
+    }
+    const expected = e1 << e2 % bitwidth;
+    cases.push({ input: [V(e1), u32(e2)], expected: V(expected) });
+  }
+  return cases;
+}
+
+// Returns all cases of shifting e1 right by [0,63]. If `is_const` is true, cases that are
+// invalid for const eval are not returned.
+function generate_shift_right_cases(e1: number, e1Type: string, is_const: boolean): CaseList {
+  const V = e1Type === 'i32' ? i32 : u32;
+  const cases: CaseList = [];
+  for (let shift = 0; shift < 64; ++shift) {
+    const e2 = shift;
+    if (is_const && !is_valid_const_shift_right(e1, e1Type, e2)) {
+      // console.log(`Skipping: ${e1} >> ${e2}`);
+      continue;
+    }
+
+    let expected: number = 0;
+    if (is_unsiged(e1Type)) {
+      // zero-fill right shift
+      expected = e1 >>> e2;
+    } else {
+      // arithmetic right shift
+      expected = e1 >> e2;
+    }
+    cases.push({ input: [V(e1), u32(e2)], expected: V(expected) });
+  }
+  return cases;
+}
+
+g.test('shift_left_concrete')
+  .specURL('https://www.w3.org/TR/WGSL/#bit-expr')
+  .desc(
+    `
+e1 << e2
+
+Shift left (shifted value is concrete)
+`
+  )
+  .params(u =>
+    u
+      .combine('type', ['i32', 'u32'] as const)
+      .combine('inputSource', allInputSources)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    const type = scalarType(t.params.type);
+    const V = t.params.type === 'i32' ? i32 : u32;
+    const is_const = t.params.inputSource === 'const';
+
+    const cases: CaseList = [
+      {
+        input: /*  */ [V(0b00000000000000000000000000000001), u32(1)],
```

Comment:
What's up with the empty comments? I assume you are trying to force formatting to align things to be easier to read.

If so you can tag a block with `// prettier-ignore` to tell the formatter to not rewrite a section of code. To be clear, it works on blocks not lines, so you can apply to to the whole array declaration at once if you want, instead of Case by Case. Took me awhile to realize the impact of the distinction that it works on blocks :upside_down_face:

Reply:
I don't like prettier-ignore, as I prefer when the formatter takes care of formatting the lines for me. This is an easy way to get both (alignment, and formatting).

Reply:
:man_shrugging: I don't feel strongly about this.
Given that you can get the formatting/alignment you want with small comments, I think this is fine. If you were having to go through more convoluted things I would be more in favour of ignoring the formatter. Mostly wanted to make you aware of the option if you were not already familiar with.

Reply:
I hear you. We've been using similar tricks in Tint for clang-format.